### PR TITLE
Configuration cache support

### PR DESCRIPTION
**Pull request by deepy**
With the exception of NodeSetupTask this adds configuration cache support.

It changes many tasks classes to be `abstract` so we can inject the various services (like `ProjectLayout`) which will break backwards compatibility. The alternatives seem to be constructor injection which doesn't look great as it bubbles upwards, or property injection but with a dummy method which I guess is the real alternative.
But with that in mind we might want to include this in 3.x and provide another RC if we get the unit-tests working again.

With that said, as was mentioned in #111 unit tests fails to capture the invocation of execSpec: [NpmInstallTaskTest.groovy#L15](https://github.com/node-gradle/gradle-node-plugin/blob/3.0.0-rc4/src/test/groovy/com/github/gradle/node/npm/task/NpmInstallTaskTest.groovy#L15)

The mock and spy is setup like this: [AbstractTaskTest.groovy#L31-L45](https://github.com/node-gradle/gradle-node-plugin/blob/3.0.0-rc4/src/test/groovy/com/github/gradle/node/task/AbstractTaskTest.groovy#L31-L45) and while this works fine with project I can't get this same pattern to work with `ExecOperations`. 
I tried the same pattern but can't find anything to hook onto, I tried retrieving the `DefaultProjectApiHelper` and changing the `execOperations` through metaprogramming but with no results.

So tests should work with `5.6.4` and `6.0` but fail on `>= 6.6.1`. We're currently using `6.8-rc-1` so that we can make use of `--included-build` and `--configuration-cache` at the same time.

### CODE REVIEW

**Comment by deepy** [comment]
Oh there's also no tests with `--configuration-cache` yet

**Comment by deepy** [comment]
There's this nice addition to `Properties` that I hadn't seen before:
```java
    /**
     * Returns a view of this {@link Provider} which can be safely read at configuration time.
     *
     * @since 6.5
     */
    @Incubating
    Provider<T> forUseAtConfigurationTime();
```

This makes the properties hilariously verbose, but if we replace some of the lazy's with the property directly from the extension it works.
e.g.

```diff
     @get:Input
-    val download by lazy { nodeExtension.download }
+    val download = nodeExtension.download.forUseAtConfigurationTime()
```

With this change the `simple-node` example project runs with no issues :-)
(* that is, none that we created. ```- Class `org.jetbrains.kotlin.daemon.common.DaemonParamsKt`: read system property 'kotlin.environment.keepalive'```)

**Comment by deepy** [comment]
Oh and for `download = true` there's also the following which I believe needs to move almost entirely to the plugin application, well at least the creation of the configuration.
```kotlin
    private fun resolveSingle(name: String): File {
        val dep = project.dependencies.create(name)
        val conf = project.configurations.detachedConfiguration(dep)
        conf.isTransitive = false
        return conf.resolve().single()
    }
```

**Comment by deepy** [comment]
@bsautel I've rebased `move-repository-creation` on top of this branch and merged it here, I'll push it in a bit, it makes nodeSetupTask a lot happier in the case of `download = false`, with the headache of the detachedConfiguration untouched.

I'm unable to run it on the kotlin examples, so I'm going to make a copy of it and turn that to groovy

**Comment by deepy** [comment]
With the latest changes the `gradle-node-plugin` part of `examples/spring-boot-angular` (when turned to groovy, skipping spring, and using `download = false`) work fine

**Comment by bsautel** [comment]
Indeed, using lazy properties here seems to be useless. And I am not sure to understand what  calling `forUseAtConfigurationTime()` changes.

What's happening with the `simple-node` or the `spring-boot-example` projects? They are working in the CI. Is it due to the merge with the repository declaration move?

I was about to suggest you to merge the repository declaration move to the configuration cache branch, that's nice you already did it! 😊

It still remains the repository usage to resolve the Node.js bundle file. I did not think about moving this code to the plugin. In this case, I assume that it means that we have to add an input to the `NodeSetupTask` which is the bundle file and the plugin sets it in the lazy configuration code of this task. But it means that the file will be downloaded at configuration time and not at task run time. Is this a big deal? It will do some unnecessary job if you declare the plugin with `download = false` and do not execute any task belonging to the plugin (since all of them depend on the `nodeSetup` task). This is maybe a satisfying solution.

**Comment by deepy** [comment]
I gave moving the creation to the plugin an attempt but it's messy, since we can't create the repository before the evaluation is complete and we have the populated `NodeExtension` that means in `afterEvaluate` we need to: create the ivy repository (or not), resolve the filename and all details of it (VariantComputer has a single method for this) and create a `detachedConfiguration` with those details, then pass *the result* from the configuration to the `NodeSetupTask`'s configuration.

I tried and messed up so I opted for looking at other things first.

As for the examples, we're not building with `--configuration-cache` and even if we were we'd only know if they passed when we ran it a second time. So when testing it locally I do: `rm -rf .gradle/configuration-cache;./gradlew --configuration-cache helloWorld;./gradlew --configuration-cache helloWorld` and the second run is the one that determines success or not.
However, the first run fails because of the kotlin DSL not being ready for the configuration cache yet :D

**Comment by bsautel** [comment]
I could successfully move the Node.js archive download to the plugin class. Thanks to the provider, it sounds like I could defer the file download in a `map` operation to the time the value is queried, i.e task execution (I did not check to ensure this is right).

I enabled the configuration cache in all the integration tests that are running in Gradle >= 6.6. They now all are successful.

I also enabled the configuration cache in the examples build and it also works. I upgraded to Spring Boot 2.4.2 (it sounds like the configuration cache support was added in the Spring Boot plugin in 2.4.0) and it works. Maybe it was not working because of the Spring Boot version of because of the remaining `project` usages.

Tell me if you have any feedback regarding these changes. I think that we can now merge this pull request to the `master` branch and maybe create a new RC version.

**Comment by deepy** [comment]
So about the examples, there's this problem with the Kotlin DSL plugin development not being ready for configuration cache :-)
https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:plugins
Although Kotlin JVM says 1.4.20 supports it, I thought I checked versions and figured that we were on a new one.

But I think we're at a point where we can merge to master, there's just some tests that needs to be written first to verify that the configuration cache works, e.g. integration tests running the tests twice to make sure that nothing explodes (except in `download = true`, which is expected to explode)

**Comment by deepy** [comment]
Oh, I was missing 6 notifications, for sure looks better! I'll have a solid look tomorrow though!

**Comment by bsautel** [comment]
I disabled the configuration cache in the examples build. You're right, the Gradle Kotlin Plugin plugin is not yet compatible with the configuration cache and we use it through the `includeBuild` mechanism. But I don't understand why those commands worked on my laptop using Ubuntu (20.04) and not in the GitHub CI also using Ubuntu (18.04). We'll enable it again when the Gradle Kotlin Plugin plugin supports configuration cache.

I also removed the compile warnings that seem to be recent, maybe since the upgrade to Gradle 6.8 that uses now Kotlin 1.4 instead of 1.3?

**Comment by deepy** [comment]
When running on my laptop it gives those warnings on the first run, but after that it's ignored until I delete the configuration-cache and I guess GA runs with a clean cache every time.

Thinking about the tests, we probably have most of it covered:
`NpmInstall` and `YarnInstallTask` runs `npmInstall`/`yarn` twice and `NpmInstall`/`YarnInstallTask` are basically a super-powered `NpmTask`/`YarnTask`.
The `NpxTask` test `'execute npx command with custom execution configuration and check up-to-date-detection'` runs `:env` twice and while I'm right now questioning why result3 has a conditional (shouldn't the task status be the same regardless?) it doesn't explode, which is good.
And `NodeTask` seems to get a similar deal in `'download specified node version and exec simple node program and check up-to-date detection'` where it's at least run three times.

So in short, we've got "full" test-coverage 🎉

**Comment by deepy** [comment]
I think we're good to go, I can merge it to master when one additional test check goes through

**Comment by bsautel** [comment]
I just come to make some additional tests and in fact the Spring Boot Angular project works but not the Simple Node one. Don't know why but it seems it has nothing to do with the Gradle Node plugin.

Regarding the behavior change in some integration tests when we enable the configuration cache, here is my understanding. We use properties to change the task configuration (for instance its environment). Without the configuration cache, the build file is executed again and the task configuration changes. But as the environment is not in the task inputs, the task is up-to-date. When we enable the configuration cache, we need to do it using the Gradle API that declares a dependency to a property, otherwise the configuration cache will not work. When we change the value of the property to change the task configuration, Gradle detects that and invalidates the configuration cache. It sounds like this makes the task no-longer up-to-date, whereas in fact its inputs will not have changed. I'm not sure about that though.

And yes, I think we have a good test coverage of configuration cache support since all the integration tests rely use it and nearly all the features are tested in integration tests. There is something to notice though: we no longer test the plugin without the configuration cache enabled on versions that support it. I don't know whether this is a problem or not. I am not going to propose to run the integration tests using a new Gradle version since the tests take so long now. It sounds like the GitHub Actions executors are much slower than they used to be.

**Comment by deepy** [comment]
Yeah the test speed has been slowly declining and the part that confuses me the most is that mac builds are so much faster than the windows ones, 2h25m compared to almost 7 hours on windows, I wonder if the parallelization of the test just isn't detecting the right amount of cores.
Although regardless of cores 7 hours is excessive.

**Comment by deepy** [comment]
macOS says it's fine, merging!
